### PR TITLE
generate code coverage reports

### LINKED ISSUE #2746: add code coverage metric
**What is the requested feature?**
Uncovered code is guaranteed to break at some point: using the code coverage metric to assess quality of the developed code can be an asset.

**Is your request related to a specific problem?**
As a concrete example, https://github.com/GEOS-DEV/GEOS/issues/2651 happens because the `-m 0` option of GEOS is untested in CI.

**Describe the solution you'd like**
Add CI generated coverage reports (and badge in README.md) to ease development.

### PR DESCRIPTION
**Pull request by tbeltzun**
This is a WIP for generating code coverage reports for the GEOS code.

Early attempt to fix https://github.com/GEOS-DEV/GEOS/issues/2746.

I don't know if I'll have time to work on this (postdoc time is limited :/), but this might be a start.

Don't hesitate to push directly in this branch to contribute if you are interested in this feature.

https://github.com/GEOS-DEV/thirdPartyLibs/pull/255

https://github.com/GEOS-DEV/LvArray/pull/312

https://github.com/GEOS-DEV/hdf5_interface/pull/12

resolves #2746

### CODE REVIEW

**Comment by wrtobin** [comment]
@tbeltzun Do you mind opening a more general issue related to this? 

This is a good start, but as you say there is more to be done to get proper coverage reports (esp. w.r.t. our GPU platforms).

File: scripts/ci_build_and_test_in_container.sh [hunk 1]
```diff
@@ -63,7 +58,8 @@ or_die python3 scripts/config-build.py \
                -ip ${GEOSX_DIR} \
                --ninja \
                -DBLT_MPI_COMMAND_APPEND='"--allow-run-as-root;--oversubscribe"' \
-               -DGEOSX_INSTALL_SCHEMA=${GEOSX_INSTALL_SCHEMA}
+               -DGEOSX_INSTALL_SCHEMA=$([[ "$*" == *--disable-schema-deployment* ]] && echo 0 || echo 1) \
+               -DENABLE_COV=$([[ "$*" == *--test-coverage* ]] && echo 1 || echo 0)
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh:62]: Is there a way to add code coverage when running the integrated tests?

Comment (rrsettgast) [reply scripts/ci_build_and_test_in_container.sh:62]: https://about.codecov.io/blog/how-to-set-up-codecov-with-c-and-github-actions/

Comment (TotoGaz) [reply scripts/ci_build_and_test_in_container.sh:62]: It's important to have this coverage from the integrated tests in our global figures: it's a strong part of our validation and coverage.

**Comment by rrsettgast** [comment]
@tbeltzun @TotoGaz this is ready for review. I would like to get it merged and start history.

File: scripts/ci_build_and_test_in_container.sh (hunk 1 above)
Comment (rrsettgast) [reply scripts/ci_build_and_test_in_container.sh]: @TotoGaz There is just a command that you execute under the premise of a coverage tests. For instance we are doing `make tests` right now, as the coverage tests....but we can run a script that executes both unit tests and integrateTests at a later time.

File: src/cmake/GeosxOptions.cmake [hunk 2]
```diff
@@ -137,6 +135,11 @@ blt_append_custom_compiler_flag( FLAGS_VAR GEOSX_NINJA_FLAGS
                                  CLANG   "-fcolor-diagnostics"
                                )
 
+# blt_append_custom_compiler_flag( FLAGS_VAR COVERAGE_FLAGS
+#                                  GNU   "-fprofile-arcs -ftest-coverage"
+#                                  CLANG "-fprofile-instr-generate -fcoverage-mapping"
+#                                )
+
```
Comment (tbeltzun) [inline src/cmake/GeosxOptions.cmake]: How do you generate coverage data without compiler flags ?

Comment (rrsettgast) [reply src/cmake/GeosxOptions.cmake]: BLT has a wrapper for coverage

File: scripts/ci_build_and_test_in_container.sh [hunk 3]
```diff
@@ -34,13 +34,15 @@ Usage: $0
...
+  --code-coverage
+      run a code build and test.
   --data-basename output.tar.gz
       If some data needs to be extracted from the build, the argument will define the tarball. Has to be a `tar.gz`.
   --exchange-dir /path/to/exchange
       Folder to share data with outside of the container.
   --host-config host-config/my_config.cmake
       The host-config. Path is relative to the root of the repository.
-  --install-dir-basename GEOS-e42ffc1
+  --install-dir-basename
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh]: `GEOS-e42ffc1` was an example to emphasize the kind of information we were expecting.
Why did you remove it? At least, we would need a placeholder to indicate we're waiting for an argument.

File: scripts/ci_build_and_test_in_container.sh [hunk 4]
```diff
@@ -64,7 +66,7 @@ exit 1
 or_die cd $(dirname $0)/..
 
 # Parsing using getopt
-args=$(or_die getopt -a -o h --long build-exe-only,cmake-build-type:,data-basename:,exchange-dir:,host-config:,install-dir-basename:,no-install-schema,no-run-unit-tests,repository:,run-integrated-tests,sccache-credentials:,test-code-style,test-documentation,help -- "$@")
+args=$(or_die getopt -a -o h --long build-exe-only,cmake-build-type:,data-basename:,exchange-dir:,host-config:,install-dir-basename:,no-install-schema,no-run-unit-tests,repository:,run-integrated-tests,code-coverage,sccache-credentials:,test-code-style,test-documentation,help -- "$@")
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh]:
```suggestion
args=$(or_die getopt -a -o h --long build-exe-only,cmake-build-type:,code-coverage,data-basename:,exchange-dir:,host-config:,install-dir-basename:,no-install-schema,no-run-unit-tests,repository:,run-integrated-tests,sccache-credentials:,test-code-style,test-documentation,help -- "$@")
```

File: scripts/ci_build_and_test_in_container.sh [hunk 5]
```diff
@@ -74,6 +76,7 @@ RUN_UNIT_TESTS=true
 RUN_INTEGRATED_TESTS=false
 TEST_CODE_STYLE=false
 TEST_DOCUMENTATION=false
+CODE_COVERAGE=false
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh:79]:
```suggestion
CODE_COVERAGE=0
```
If you directly set `CODE_COVERAGE` to `0` or `1` (line 107) then you won't have to use the fancy `-DENABLE_COVERAGE=$([[ "${CODE_COVERAGE}" = true ]] && echo 1 || echo 0)` and you'll be able to use `CODE_COVERAGE` directly.

File: scripts/ci_build_and_test_in_container.sh [hunk 6]
```diff
@@ -207,6 +219,16 @@ else
...
 fi
 
+if [[ ! -z "${SCCACHE_CREDS}" ]]; then
+  echo "sccache post-build state"
+  or_die ${SCCACHE} --show-adv-stats
+fi
+
+if [[ "${CODE_COVERAGE}" = true ]]; then
+  or_die ninja coreComponents_coverage
+  cp -r ${GEOSX_BUILD_DIR}/coreComponents_coverage.info.cleaned /tmp/geos/geos_coverage.info.cleaned
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh]:
```suggestion
  cp -r ${GEOSX_BUILD_DIR}/coreComponents_coverage.info.cleaned ${GEOS_SRC_DIR}/geos_coverage.info.cleaned
```
?

Or maybe using `DATA_EXCHANGE_DIR` if it does not interfere with the integrated tests (not sure).

File: src/coreComponents/CMakeLists.txt [hunk 7]
```diff
@@ -108,7 +108,7 @@ endif()
 
 geosx_add_code_checks( PREFIX coreComponents
                        EXCLUDES cmake
-                       src/coreComponents/constitutive/PVTPackage )
+                                constitutive/PVTPackage )
```
Comment (TotoGaz) [inline src/coreComponents/CMakeLists.txt]: Warning about the indentation.

File: src/coreComponents/dataRepository/CMakeLists.txt [hunk 8]
```diff
@@ -59,8 +59,6 @@ blt_add_library( NAME       dataRepository
 
 target_include_directories( dataRepository PUBLIC ${CMAKE_SOURCE_DIR}/coreComponents )
 
-geosx_add_code_checks( PREFIX dataRepository )
```
Comment (TotoGaz) [inline src/coreComponents/dataRepository/CMakeLists.txt:62]: Why did you remove all of these calls?

File: src/cmake/GeosxOptions.cmake [hunk 9]
```diff
@@ -137,6 +135,11 @@ blt_append_custom_compiler_flag( FLAGS_VAR GEOSX_NINJA_FLAGS
                                  CLANG   "-fcolor-diagnostics"
                                )
 
+# blt_append_custom_compiler_flag( FLAGS_VAR COVERAGE_FLAGS
+#                                  GNU   "-fprofile-arcs -ftest-coverage"
+#                                  CLANG "-fprofile-instr-generate -fcoverage-mapping"
+#                                )
```
Comment (TotoGaz) [inline src/cmake/GeosxOptions.cmake]: Do you want to remove this eventually?

File: scripts/ci_build_and_test_in_container.sh [hunk 10]
```diff
@@ -153,6 +157,13 @@ if [[ "${RUN_INTEGRATED_TESTS}" = true ]]; then
   ATS_CMAKE_ARGS="-DATS_ARGUMENTS=\"--machine openmpi --ats openmpi_mpirun=/usr/bin/mpirun --ats openmpi_args=--allow-run-as-root --ats openmpi_procspernode=2 --ats openmpi_maxprocs=2\" -DPython3_ROOT_DIR=${ATS_PYTHON_HOME}"
 fi
 
+
+if [[ "${CODE_COVERAGE}" = true ]]; then
+  apt-get install -y lcov
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh]:
```suggestion
  or_die apt-get update
  or_die apt-get install -y lcov
```
You may need to update the cache of packages before you try to install your package.

File: scripts/ci_build_and_test_in_container.sh [hunk 11]
```diff
@@ -207,6 +219,16 @@ else
   fi
 fi
 
+if [[ ! -z "${SCCACHE_CREDS}" ]]; then
+  echo "sccache post-build state"
+  or_die ${SCCACHE} --show-adv-stats
+fi
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh:226]: Maybe should we remove the `final state` lines 260-263? It's either duplicated or "false" (with code coverage disturbing the process for whatever reason)?

File: src/cmake/GeosxOptions.cmake [hunk 12]
```diff
@@ -122,8 +120,8 @@ message( "CMAKE_CXX_COMPILER_ID = ${CMAKE_CXX_COMPILER_ID}" )
 
 blt_append_custom_compiler_flag( FLAGS_VAR CMAKE_CXX_FLAGS DEFAULT "${OpenMP_CXX_FLAGS}" )
 blt_append_custom_compiler_flag( FLAGS_VAR CMAKE_CXX_FLAGS
-                                 GNU   "-Wall -Wextra -Wpedantic -pedantic-errors -Wshadow -Wfloat-equal -Wcast-align -Wcast-qual"
-                                 CLANG "-Wall -Wextra -Wpedantic -pedantic-errors -Wshadow -Wfloat-equal -Wno-cast-align -Wcast-qual"
+                                 GNU   "-Wpedantic -pedantic-errors -Wshadow -Wfloat-equal -Wcast-align -Wcast-qual"
+                                 CLANG "-Wpedantic -pedantic-errors -Wshadow -Wfloat-equal -Wno-cast-align -Wcast-qual"
```
Comment (TotoGaz) [inline src/cmake/GeosxOptions.cmake:124]: Does this mean that you are completely removing those flags or were they duplicated?

File: scripts/ci_build_and_test_in_container.sh [hunk 13]
```diff
@@ -174,7 +185,8 @@ or_die python3 scripts/config-build.py \
                -ip ${GEOSX_DIR} \
                --ninja \
                -DBLT_MPI_COMMAND_APPEND='"--allow-run-as-root;--oversubscribe"' \
-               -DGEOSX_INSTALL_SCHEMA=${GEOSX_INSTALL_SCHEMA} \
+               -DGEOSX_INSTALL_SCHEMA=$([[ "$*" == *--disable-schema-deployment* ]] && echo 0 || echo 1) \
```
Comment (TotoGaz) [inline scripts/ci_build_and_test_in_container.sh]: We normally have the `--no-install-schema` option that populates the `GEOSX_INSTALL_SCHEMA` variable.
Why lead you into this change?

File: src/cmake/GeosxMacros.cmake [hunk 14]
```diff
@@ -53,6 +53,12 @@ macro( geosx_add_code_checks )
                              UNCRUSTIFY_CFG_FILE ${PROJECT_SOURCE_DIR}/uncrustify.cfg )
     endif()
 
+    if (ENABLE_COVERAGE)
+        blt_add_code_coverage_target(NAME   ${arg_PREFIX}_coverage
+                                    RUNNER ctest -E 'blt_gtest_smoke|blt_mpi_smoke|testUncrustifyCheck|testDoxygenCheck'
+                                    SOURCE_DIRECTORIES ${PROJECT_SOURCE_DIR}/coreComponents )
+    endif()
```
Comment (TotoGaz) [inline src/cmake/GeosxMacros.cmake]: Warning source code formatting.

**Review by TotoGaz — COMMENTED** [review]
a few comments

File: scripts/ci_build_and_test_in_container.sh (hunk 3 above)
Comment (rrsettgast) [reply scripts/ci_build_and_test_in_container.sh]: I thought I had accidentally pasted that in...haha.

File: src/cmake/GeosxOptions.cmake (hunk 12 above)
Comment (rrsettgast) [reply src/cmake/GeosxOptions.cmake:124]: Yes completely removing these flags here because they were duplicated in BLT.

File: scripts/ci_build_and_test_in_container.sh (hunk 13 above)
Comment (rrsettgast) [reply scripts/ci_build_and_test_in_container.sh]: I didn't. This was here from @tbeltzun

File: scripts/ci_build_and_test_in_container.sh (hunk 11 above)
Comment (rrsettgast) [reply scripts/ci_build_and_test_in_container.sh:226]: yes. I agree.

File: scripts/ci_build_and_test_in_container.sh (hunk 6 above)
Comment (rrsettgast) [reply scripts/ci_build_and_test_in_container.sh]: I meant to do this but forgot. thanks.

File: src/cmake/GeosxOptions.cmake (hunk 2 above)
Comment (rrsettgast) [inline src/cmake/GeosxOptions.cmake]:
```suggestion
```

File: src/coreComponents/dataRepository/CMakeLists.txt (hunk 8 above)
Comment (rrsettgast) [reply src/coreComponents/dataRepository/CMakeLists.txt:62]: I think they are unnecessary to create all these component level targets.

File: src/cmake/GeosxOptions.cmake [hunk 15]
```diff
@@ -150,6 +148,10 @@ if( ${CMAKE_MAKE_PROGRAM} STREQUAL "ninja" OR ${CMAKE_MAKE_PROGRAM} MATCHES ".*/
   set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${GEOSX_NINJA_FLAGS}" )
 endif()
 
+if( ENABLE_COV )
```
Comment (wrtobin) [inline src/cmake/GeosxOptions.cmake]: I'm seeing `ENABLE_COV` and `ENABLE_COVERAGE` in different cmake files, are they different or intended to be the same/synonymous?

**Review by wrtobin — APPROVED** [review]
Except for the `ENABLE_COV/COVERAGE` note I had, this looks good.

File: src/cmake/GeosxOptions.cmake (hunk 15 above)
Comment (rrsettgast) [reply src/cmake/GeosxOptions.cmake]: not intended
